src/rheology/Viscosity.jl

```diff
@@ -30,32 +21,109 @@ end
         # compute and update stress viscosity
         ηi = compute_viscosity_εII(rheology, εII, args_ij)
         ηi = continuation_log(ηi, η[i, j], ν)
-        η[i, j] = clamp(ηi, 1e16, 1e24)
+        η[i, j] = clamp(ηi, cutoff...)
     end
 
     return nothing
 end
 
-@parallel_indices (i, j) function compute_viscosity!(η, ν, εII, args, rheology)
+@parallel_indices (i, j) function compute_viscosity!(η, ν, εII, args, rheology, cutoff)
     @inbounds begin
         # argument fields at local index
-        args_ij = local_args(args, i, j)
+        args_ij = local_viscosity_args(args, i, j)
 
         # compute second invariant of strain rate tensor
         εII_ij = εII[i, j]
 
         # compute and update stress viscosity
         ηi = compute_viscosity_εII(rheology, εII_ij, args_ij)
         ηi = continuation_log(ηi, η[i, j], ν)
-        η[i, j] = clamp(ηi, 1e16, 1e24)
+        η[i, j] = clamp(ηi, cutoff...)
     end
 
     return nothing
 end
 
-#3D kernel
+function _compute_viscosity!(η, ν, ratios_center, εxx, εyy, εxyv, args, rheology, cutoff)
+    I = JustRelax.indices(size(η))
+
+    # convinience closure
+    @inline gather(A) = _gather(A, I...)
+
+    if all(I .<= size(η))
+        # we need strain rate not to be zero, otherwise we get NaNs
+        εII_0 = (εxx[I...] == εyy[I...] == 0) * 1e-15
+
+        # argument fields at local index
+        args_ij = JustRelax.local_viscosity_args(args, I...)
+
+        # local phase ratio
+        ratio_ij = ratios_center[I...]
+
+        # compute second invariant of strain rate tensor
+        εij = εII_0 + εxx[I...], -εII_0 + εyy[I...], gather(εxyv)
+        εII = second_invariant(εij...)
+
+        # compute and update stress viscosity
+        ηi = JustRelax.compute_phase_viscosity_εII(rheology, ratio_ij, εII, args_ij)
+        ηi = JustRelax.continuation_log(ηi, η[I...], ν)
+        η[I...] = clamp(ηi, cutoff...)
+    end
+
+    return nothing
+end
+
+function compute_viscosity!(η::CuArray{T1,N,T2}, x...) where {T1,T2,N}
```

Comment:
testing or did you mean to add this with CuArray specifically?

Reply:
Was testing, and indeed this change should be reversed